[CAD-2671] Modify cardano-sync to enable reusing from SMASH.

ksaric opened the pull request:
https://jira.iohk.io/browse/CAD-2671

erikd commented on cardano-sync/src/Cardano/Sync.hs:
```diff
@@ -146,8 +152,15 @@ runSyncNode dataLayer plugin enp insertValidateGenesisDist =
             syncEnv <- ExceptT $ mkDbSyncEnvFromConfig dataLayer (enpLedgerStateDir enp) genCfg
-            liftIO $ runDbSyncNodeNodeClient (dncPrometheusPort enc) syncEnv
-                iomgr trce plugin (cardanoCodecConfig bCfg) (enpSocketPath enp)
+            liftIO $ runDbSyncNodeNodeClient
```
Why change the layout style from what was there. A good rule of thumb when submitting a PR to a project is to follow the existing coding style, eg "When in Rome do as the Romans".

ksaric replied:
Fine, I can revert that.

erikd commented on cardano-db/src/Cardano/Db/Schema/Types.hs:
```diff
@@ -8,6 +8,8 @@ module Cardano.Db.Schema.Types
   ) where
 
+import           Cardano.Prelude
+
```
What benefit does this actually give us?

ksaric replied:
It's a standard prelude we are supposed to use? I guess I don't have to talk about the benefits of a custom Prelude?

erikd replied:
> It's a standard prelude we are supposed to use? 

No, there is no mandate.

erikd commented on cardano-db/src/Cardano/Db/Migration/Version.hs:
```diff
@@ -23,9 +23,9 @@ data MigrationVersion = MigrationVersion
 
 
-parseMigrationVersionFromFile :: String -> Maybe MigrationVersion
+parseMigrationVersionFromFile :: Text -> Maybe MigrationVersion
```
Why are these types changing from `String` to `Text`? It does not seem to be related to "Modify cardano-sync to enable reusing from SMASH:.

ksaric replied:
Well. 
First, I noticed that it would be better if I add `cardano-prelude` in a single module.
Then, there was a collision in that module between `cardano-prelude` and base Prelude.
Then I added the extension for the whole library which I consider worthy, `NoImplicitPrelude`.
Then other modules started complaining.
Then I fixed the other modules.

After a while, I realized it might have been better to do this in another PR, but you don't know how extensive the changes are until you do them.

erikd commented on cardano-db/src/Cardano/Db/Migration/Version.hs:
```diff
@@ -40,18 +40,18 @@ nextMigrationVersion (MigrationVersion _stage ver _date) = do
+renderMigrationVersion :: MigrationVersion -> Text
 renderMigrationVersion mv =
-  List.intercalate "-"
+  toS $ intercalate "-"
```
Why are we moving from qualified ` List.intercalate` to unqualified `intercalate` ? I know `Cardano.Prelude` exposes `intercalate` but I do not see a good reason to switch from qualified to unqualified usage.

ksaric replied:
https://github.com/input-output-hk/cardano-db-sync/pull/527#discussion_r575045248

erikd commented on cardano-db/cardano-db.cabal:
```diff
@@ -20,6 +20,9 @@ extra-source-files:     CHANGELOG.md
   default-language:     Haskell2010
+  default-extensions:
+    NoImplicitPrelude
+    OverloadedStrings
```
This seems un-related to the aim of the PR. In general I think defining default extensions in the cabal file is a bad idea. Just about the only one where I think this is not a bad idea is `NoImplicitPrelude` .

ksaric replied:
https://github.com/input-output-hk/cardano-db-sync/pull/527#discussion_r575045248

ksaric replied:
I think that using `Text` is what you want to do, in most cases, so this is something that I used quite a lot. Additionally, the current `cardano-prelude` does not export `String`. So there is that.

erikd commented on cardano-db/src/Cardano/Db/Query.hs:
```diff
@@ -65,6 +65,7 @@ module Cardano.Db.Query
   , unValueSumAda
   ) where
 
+import           Cardano.Prelude hiding (Meta, from, isJust, isNothing, maybeToEither, on)
```
This change does not offer a net benefit.

ksaric replied:
https://github.com/input-output-hk/cardano-db-sync/pull/527#discussion_r575045248

erikd reviewed (changes requested):
* Rebase against master.
* The change that includes `Cardano.Prelude` everywhere needs to be split out into a separate commit.
* Remove language pragmas from cabal files and put them in the files that need then (only).
* Fix all un-used import warnings (due ot use of `Cardano.Prelude`).

ksaric commented:
I rebased against `master` and reverted most of the changes related to the `cardano-prelude` and removed the excess pragmas. Should be ok for now, we can always propagate it through the codebase in later PRs, if so required.
